www/docs/server/validators.md

```diff
@@ -0,0 +1,343 @@
+---
+id: validators
+title: Input & Output Validators
+sidebar_label: Input & Output Validators
+slug: /server/validators
+---
+
+tRPC procedures may define validation logic for their input and/or output. There is first class support for many popular validators, and [integrating a validator](#contributing-your-own-validator-library) which we don't directly support in usually straight-forward.
+
+### Input Validators
+
+By defining an input validator, tRPC can check that a procedure call is correct and return a validation error if not. 
+
+To set up an input validator, use the `procedure.input()` method:
+
+```ts twoslash
+// @target: esnext
+import { initTRPC } from '@trpc/server';
+import { z } from 'zod';
+
+export const t = initTRPC.create();
+const publicProcedure = t.procedure
+
+// ---cut---
+
+export const appRouter = t.router({
+  hello: publicProcedure
+    .input(
+      z.object({
+        name: z.string(),
+      }),
+    )
+    .query((opts) => {
+      const name = opts.input.name;
+      //    ^?
+      return {
+        greeting: `Hello ${opts.input.name}`,
+      };
+    }),
+});
+```
+
+`.input` can also be stacked to build more complex types, which is particularly useful when you want to utilise some common input to a collection of procedures in a [middleware](middlewares).
+
+```ts twoslash
+// @target: esnext
+import { TRPCError, initTRPC } from '@trpc/server';
+import { z } from 'zod';
+
+export const t = initTRPC.create();
+
+// ---cut---
+
+const procedureWhichRejectsOutOfTownFolk = t.procedure
+  .input(z.object({ townName: z.string() }))
+  .use((opts) => {
+    if (opts.input.townName !== 'Pucklechurch') {
+      throw new TRPCError({
+        code: 'FORBIDDEN',
+        message: "We don't take kindly to out-of-town folk",
+      });
+    }
+
+    return opts.next();
+  });
+
+export const appRouter = t.router({
+  hello: procedureWhichRejectsOutOfTownFolk
+    .input(
+      z.object({
+        name: z.string(),
+      }),
+    )
+    .query((opts) => {
+      const input = opts.input;
+      //    ^?
+      return {
+        greeting: `Hello ${input.name}, my friend`,
+      };
+    }),
+});
+```
+
+### Output Validators
+
+By defining an output validator, tRPC can also check that data returned from untrusted sources is correct, or to strip extraneous data. Defining outputs is not always as important as defining inputs, since tRPC gives you automatic type-safety by inferring the return type of your procedures.
```

Comment:
```suggestion
You can also define an output validator. Validating outputs is not always as important as defining inputs, since tRPC gives you automatic type-safety by inferring the return type of your procedures. Some reasons to define an output validator include:

- Checking that data returned from untrusted sources is correct
- Ensure that you are not returning more data to the client than necessary
```